### PR TITLE
ARROW-9977: [Rust] Added min/max of [Large]StringArray

### PR DESCRIPTION
**Pull request by jorgecarleitao**
Like the title says, another helper.

Note that, contrarily to horizontal operations, that we can represent as `f(&ArrayRef) -> ArrayRef`, in vertical operations like `min`, and `sum` we do not have a dynamically-typed `struct` to return, and thus need to have a function signature for each type. :(

I wonder if we could have in Rust a dynamically struct for single values, like DataFusion has `ScalarValue`.

### CODE REVIEW

**Comment by nevi-me** [comment]
There's a lot of value in creating StructValue in arrow, like cpp has. One of the potential benefits would be showing the compute kennels to either take an array or scalar

I've tried doing this previously, but threw away the work as I couldn't get it quite right.

**Comment by jorgecarleitao** [comment]
@nevi-me , to understand, the goal would we a scalar that we could safely convert `Vec<Scalar>` from and to an array? What is the issue with

```
enum ScalarValue {
    Int32(Option<i32>)
    ...
}
```

**Comment by andygrove** [comment]
I think that @nevi-me 's suggestion addresses something that I've been trying to solve as well. Our kernels operate on arrays, so calling `a + b` is fine when both are arrays, but if we're evaluating `a + 1` against a batch of 1000 rows then we have to create an array of 1000 literal ones to pass to the kernel.

File: rust/arrow/src/compute/kernels/aggregate.rs [hunk 1]
```diff
@@ -19,9 +19,42 @@
...
+                let item = $array.value(i);
+                if !has_value || (&n $cmp &item) {
+                    has_value = true;
+                    n = item;
+                }
+            }
+        } else {
+            for i in 0..data.len() {
+                let item = $array.value(i);
+                if !has_value || data.is_valid(i) && (&n $cmp &item) {
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/aggregate.rs:48]: Is this correct? `&&` has higher precedence than `||` so I am reading this as `if !has_value || (data.is_valid(i) && (&n $cmp &item))` and since `!has_value` is true on the first iteration of the loop, it will always set `has_value = true` ?

File: rust/arrow/src/compute/kernels/aggregate.rs [hunk 2]
```diff
@@ -149,4 +202,18 @@ mod tests {
...
+
+    #[test]
+    fn test_string_min_max_with_nulls() {
+        let a = StringArray::from(vec![Some("b"), None, None, Some("a"), Some("c")]);
+        assert_eq!("a", min_string(&a).unwrap());
+        assert_eq!("c", max_string(&a).unwrap());
+    }
+
+    #[test]
+    fn test_string_min_max_all_nulls() {
```
Comment (andygrove) [inline rust/arrow/src/compute/kernels/aggregate.rs:214]: Could you add an additional test with an array with one or more `None` at the start, followed by one or more `Some` to see if I am correct about my earlier comment?

File: rust/arrow/src/compute/kernels/aggregate.rs (hunk 1 above)
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/aggregate.rs]: You are right. Amazing catch 💯💯.  I pushed a fix to it. This was also the case for primitive types, and so I also added a test and fix to them.

File: rust/arrow/src/compute/kernels/aggregate.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/aggregate.rs:221]: done

**Review by andygrove — APPROVED** [review]
LGTM
